### PR TITLE
Add hook support

### PR DESCRIPTION
**Pull request by rcoelho**
Hi @rs  I improved my implementation (#24), making it faster and removing all allocations.

Since this adds a change to the Event struct (adds the hasLevel field), I'm sending a separate PR.
Please check (and hopefully merge) #24 before this one.

Thanks!

### CODE REVIEW

**Comment by rcoelho** [comment]
Some benchmarks:

Before the improvements:

```
BenchmarkHooks/Nop/Single-8             20000000                56.3 ns/op            40 B/op          2 allocs/op
BenchmarkHooks/Nop/Multi-8              20000000                81.8 ns/op            80 B/op          3 allocs/op
BenchmarkHooks/Simple-8                 20000000                91.3 ns/op            40 B/op          2 allocs/op
```

After the first commit:

```
BenchmarkHooks/Nop/Single-8             30000000                35.4 ns/op            16 B/op          1 allocs/op
BenchmarkHooks/Nop/Multi-8              30000000                45.1 ns/op            32 B/op          1 allocs/op
BenchmarkHooks/Simple-8                 30000000                56.6 ns/op            16 B/op          1 allocs/op
```

Aftere the second commit:

```
BenchmarkHooks/Nop/Single-8             50000000                22.4 ns/op             0 B/op          0 allocs/op
BenchmarkHooks/Nop/Multi-8              50000000                25.8 ns/op             0 B/op          0 allocs/op
BenchmarkHooks/Simple-8                 30000000                42.5 ns/op             0 B/op          0 allocs/op
```

File: event.go [hunk 1]
```diff
@@ -21,10 +21,12 @@ var eventPool = &sync.Pool{
...
-	buf   []byte
-	w     LevelWriter
-	level Level
-	done  func(msg string)
+	buf      []byte
+	w        LevelWriter
+	level    Level
+	done     func(msg string)
+	h        []Hook
+	hasLevel bool
```
Comment (rs) [inline event.go]: What about adding a new `NoLevel` level before `DebugLevel` in the enum so we don't need this extra field?

File: event.go [hunk 2]
```diff
@@ -75,6 +78,14 @@ func (e *Event) Msg(msg string) {
...
 		return
 	}
+	if len(e.h) > 0 {
+		e.h[0].Run(e, e.level, e.hasLevel, msg)
+		if len(e.h) > 1 {
+			for _, hook := range e.h[1:] {
+				hook.Run(e, e.level, e.hasLevel, msg)
+			}
+		}
+	}
```
Comment (rs) [inline event.go]: Could you factorize this so you don't repeat the same code in `Msgf`?

File: log.go [hunk 3]
```diff
@@ -155,6 +173,10 @@ func (l Logger) Output(w io.Writer) Logger {
 	l2 := New(w)
 	l2.level = l.level
 	l2.sampler = l.sampler
+	if len(l.hooks) > 0 {
+		l2.hooks = make([]Hook, len(l.hooks), cap(l.hooks))
```
Comment (rs) [inline log.go]: You don't need to set the cap if equal to the size.

File: log.go [hunk 4]
```diff
@@ -155,6 +173,10 @@ func (l Logger) Output(w io.Writer) Logger {
...
+		copy(l2.hooks, l.hooks)
```
Comment (rs) [inline log.go]: `append` should be slightly faster in this context (with len=0 on prev line indeed).

**Comment by rs** [comment]
Do you think we could leverage the `Event.done` property which is already executed in `Msg` and `Msgf` to implement this?

File: event.go (hunk 1 above)
Comment (rcoelho) [reply event.go]: I thought about doing it, but wanted to change as little as possible. I'll send some more commits your way.

Comment (rs) [reply event.go]: It would let us remove the ugly `addLevelField` param too.

Comment (rcoelho) [reply event.go]: This part is done. The code is simpler now (benchmarks say it is faster, too).
After the merge could you please handle the NoLog case in console.go?
Please take a look at syslog.go, too.

File: log.go [hunk 5]
```diff
@@ -274,9 +308,7 @@ func (l *Logger) WithLevel(level Level) *Event {
 //
 // You must call Msg on the returned event in order to send the event.
 func (l *Logger) Log() *Event {
```
Comment (rs) [inline log.go]: The description if this method is no longer true.

Comment (rcoelho) [reply log.go]: Sorry, I didn't understand this comment. It is still necessary to call Msg on the returned event.
Or are you referencing the comment "We use panic level..."? It's removed now.
Or something else entirely?

Comment (rs) [reply log.go]: Never mind, I read too fast.

**Comment by rs** [comment]
Will you address other comments in a subsequent commit?

**Comment by rcoelho** [comment]
I will, but not today. Thanks for the reviews!

File: event.go (hunk 2 above)
Comment (rcoelho) [reply event.go]: Event.Msgf

File: log.go (hunk 3 above)
Comment (rcoelho) [reply log.go]: I'm just appending now.

File: log.go (hunk 4 above)
Comment (rcoelho) [reply log.go]: Yeah, using append now, and without using make first.

**Comment by rcoelho** [comment]
> Do you think we could leverage the Event.done property which is already executed in Msg and Msgf to implement this?

Possible, but would complicate things. I think that it is good the way it is now.
Ready to merge if all changes look good to you.

**Pull request merged**